apps/fz_http/lib/fz_http/rules.ex

```diff
@@ -51,28 +61,69 @@ defmodule FzHttp.Rules do
     )
   end
 
+  def nftables_device_spec(device) do
+    rules = FzHttp.Rules.list_rules(device.user_id)
+
+    Enum.map(rules, fn rule ->
+      {get_device_ip(device, ip_type("#{rule.destination}")), decode(rule.destination),
+       rule.action}
+    end)
+    # XXX: This should only be needed if a destination can be ipv4/v6 while ipv4/v6
+    # is disabled in firezone
+    |> Enum.reject(fn {source, _, _} -> is_nil(source) end)
+    |> Enum.map(fn {s, d, r} -> {decode(s), d, r} end)
+  end
+
   def nftables_spec(rule) do
-    {decode(rule.destination), rule.action}
+    nftables_spec(rule, rule.user_id)
+  end
+
+  defp nftables_spec(rule, nil) do
+    [{decode(rule.destination), rule.action}]
+  end
+
+  defp nftables_spec(rule, user_id) do
+    get_user_ips(user_id, ip_type("#{rule.destination}"))
+    # XXX: This should only be needed if a destination can be ipv4/v6 while ipv4/v6
+    # is disabled in firezone
+    |> Enum.reject(fn ip -> is_nil(ip) end)
+    |> Enum.map(fn source -> {decode(source), decode(rule.destination), rule.action} end)
+  end
+
+  defp get_device_ip(device, "IPv4"), do: device.ipv4
+  defp get_device_ip(device, "IPv6"), do: device.ipv6
+  defp get_device_ip(_, "unknown"), do: raise("Unknown protocol")
+
+  defp get_user_ips(user_id, "IPv4") do
+    Enum.map(FzHttp.Devices.list_devices(user_id), fn device ->
+      device.ipv4
+    end)
+  end
+
+  defp get_user_ips(user_id, "IPv6") do
+    Enum.map(FzHttp.Devices.list_devices(user_id), fn device ->
+      device.ipv6
+    end)
+  end
+
+  defp get_user_ips(_, "unknown") do
+    raise "Unknown protocol."
   end
 
   def to_nftables do
-    Enum.map(nftables_query(), fn {dest, act} ->
-      {decode(dest), act}
+    Enum.flat_map(nftables_query(), fn rule ->
```

Comment:
Hmm, do you think it'd be possible to fetch all this in one query using perhaps a [many to many](https://hexdocs.pm/ecto/Ecto.Association.ManyToMany.html) association?

```sql
SELECT rules.*, devices.ipv4 AS source_ipv4, devices.ipv6 AS source_ipv6
FROM rules
LEFT JOIN devices ON rules.user_id = devices.user_id
```

Reply:
Oh good, I missed this. [Here](https://github.com/firezone/firezone/pull/722/commits/febfb6a9253d25851f281cf78a446d83cc0d3ae0) fixed. This will probably change a lot in firezone/product#398 but this optimization or something similar will remain.